# Promise

## 1. 프로미스란?

자바스크립트는 비동기 처리를 위한 하나의 패턴으로 **콜백 함수**를 사용

하지만 전통적인 콜백 패턴은

- 가독성이 나쁘고 
- 비동기 처리 중 발생한 **에러의 예외 처리가 곤란**
- 여러 개의 비동기 처리 로직을 한꺼번에 처리하는 것도 한계가 있음

ES6에서 비동기 처리를 위해 도입된 또 다른 패턴이 **Promise**이다. Promise는 전통적인 콜백 패턴이 가진 단점을 보완하며 **비동기 처리 시점을 명확하게 표**현함

## 2. 콜백 패턴의 단점

### 2.1. 콜백 헬

> ### 동기식 처리 모델 & 비동기식 처리 모델
>
> **동기식 처리 모델(Synchronous processing model)**은 직렬적으로 task를 수행. 즉, 작업은 순차적으로 실행되며 어떤 작업이 수행 중이면 다음 작업은 **대기**하게 된다. 서버에서 데이터를 가져와 표시하는 작업을 수행할 때, 서버에 요청한 데이터가 응답될 때까지 이후의 작업들은 blocking된다.
>
> ![synchronous](https://poiemaweb.com/img/synchronous.png)
>
> **비동기식 처리 모델(Asyncronous processing model / Non-blocking processing model)**은 병렬적으로 task를 수행. 즉, 작업이 종료되지 않은 상태여도 **대기하지 않고 즉시 다음 작업을 실행**한다. 서버에서 데이터를 요청한 후 응답을 기다리지 않고(Non-blocking) 바로 다음 작업을 수행한다. 이후 서버로부터 데이터가 응답되면 이벤트가 발생하고, 이벤트 핸들러가 데이터를 가지고 다음 작업을 수행한다. 자바스크립트의 대부분의 DOM 이벤트와 Timer 함수, Ajax 요청은 비동기식 처리 모델로 동작한다.
>
> ![asynchronous](https://poiemaweb.com/img/asynchronous.png)

비동기식 처리 모델은 요청을 병렬로 처리해 다른 요청이 blocking되지 않는 장점이 있다.

그러나 비동기 처리를 위해 콜백 패턴을 사용하면 처리 순서를 보장하기 위해서 여러 개의 콜백 함수가 중첩되어 복잡도가 높아지는 **콜백 헬**이 발생한다.

![callback hell](https://poiemaweb.com/img/callback-hell.png)

### 2.2. 에러 처리의 한계

콜백 방식의 가장 심각한 단점은 에러 처리가 곤란하다는 것이다.

```javascript
try {
  setTimeout(() => { throw new Error('Error!'); }, 1000);
} catch (e) {
  console.log('에러를 캐치하지 못한다..');
  console.log(e);
}
```

try 블록 내에서 setTimeout이 실행되면 1초 후에 콜백 함수가 실행되면서 예외를 발생시킨다. 하지만 이 예외는 catch 블록에서 캐치되지 않는다.

비동기 처리 함수의 콜백 함수는 **해당 이벤트가 발생하면 Task Queue(Event Queue)로 이동**한 후, Call Stack이 비어졌을 때 Event Loop에 의해서 Call Stack으로 이동되어 실행한다.

setTimeout 함수는 비동기 함수이므로 콜백 함수가 실행되는 것을 기다리지 않고 바로 종료되어 Call Stack에서 제거된다. 즉, setTimeout 함수의 콜백 함수를 setTimeout 함수가 호출하지 않는 것이 문제가 된다.

예외(exception)은 호출자(caller) 방향으로 전파된다. 그러나 **비동기 처리 함수의 콜백 함수는 비동기 처리 함수에 의해 실행되지 않기 때문에** 예외 를 캐치하기 어렵다.

이런 문제를 극복하기 위해 Promise가 제안, IE를 제외한 대부분의 브라우저가 지원.

## 3. Promise 생성

프로미스는 Promise 생성자 함수를 통해 instance화 한다.

Promise 생성자 함수는 비동기 작업을 수행할 콜백 함수를 인자로 전달받음. 이 콜백 함수는 resolve와 reject 함수를 인자로 전달받음

 ```javascript
// Promise 객체의 생성
const promise = new Promise((resolve, reject) => {
  // 비동기 작업을 수행한다.

  if (/* 비동기 작업 수행 성공 */) {
    resolve('result');
  }
  else { /* 비동기 작업 수행 실패 */
    reject('failure reason');
  }
});
 ```

Promise는 비동기 처리가 성공(fulfilled)했는지 또는 실패(rejected)했는지 등의 상태 정보를 가짐

| 상태          | 의미                                       | 구현                                               |
| :------------ | :----------------------------------------- | :------------------------------------------------- |
| pending       | 비동기 처리가 아직 수행되지 않은 상태      | resolve 또는 reject 함수가 아직 호출되지 않은 상태 |
| **fulfilled** | 비동기 처리가 수행된 상태 (성공)           | resolve 함수가 호출된 상태                         |
| **rejected**  | 비동기 처리가 수행된 상태 (실패)           | reject 함수가 호출된 상태                          |
| settled       | 비동기 처리가 수행된 상태 (성공 또는 실패) | resolve 또는 reject 함수가 호출된 상태             |

```javascript
const promiseAjax = (method, url, payload) => {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.open(method, url);
    xhr.setRequestHeader('Content-type', 'application/json');
    xhr.send(JSON.stringify(payload)); // 비동기 처리

    xhr.onreadystatechange = function () {
      // 서버 응답 완료가 아니면 무시
      if (xhr.readyState !== XMLHttpRequest.DONE) return;

      if (xhr.status >= 200 && xhr.status < 400) {
        // resolve 메소드를 호출하면서 처리 결과를 전달
        resolve(xhr.response); // Success!
      } else {
        // reject 메소드를 호출하면서 에러 메시지를 전달
        reject(new Error(xhr.status)); // Failed...
      }
    };
  });
};
```

resolve와 reject 함수의 처리 결과는 Promise 객체의 후속 처리 메소드로 전달된다.

## 4. Promise의 후속 처리 메소드

Promise로 구현된 비동기 함수는 Promise 객체를 반환해야 한다. Promise로 구현된 비동기 함수를 호출하는 측(promise consumer)에서는 Promise 객체의 후속 처리 메소드(then, catch)를 통해 비동기 처리 결과 또는 에러 메시지를 전달받아 처리한다.

Promise 객체가 가지는 상태 값에 따라 후속 처리 메소드를 chaining 방식으로 호출한다.

> ### then
>
> then 메소드는 2개의 콜백 함수를 인자로 전달받음. 첫 번째는 성공(fulfilled, resolve 함수가 호출된 상태)시 호출, 두 번째는 실패(rejected, reject 함수가 호출된 상태) 시 호출됨.
>
> **then 메소드는 Promise를 반환**

> ### catch
>
> 예외(비동기 처리에서 발생한 에러와 then 메소드에서 발생한 에러)가 발생하면 호출됨.
>
> **catch 메소드는 Promise를 반환**

```javascript
asyncThing1()
    .then(function() { return asyncThing2();})
    .then(function() { return asyncThing3();})
    .catch(function(err) { return asyncRecovery1();})
 
    .then(function() { return asyncThing4();}, function(err) { return asyncRecovery2(); })
    .catch(function(err) { console.log("Don't worry about it");})
 
    .then(function() { console.log("All done!");});
```

![img](https://t1.daumcdn.net/cfile/tistory/99E159505A6F341327)

## 5. Promise의 에러 처리

then의 두 번째 콜백 함수로도 에러를 처리할 수 있고 catch를 이용할 수도 있다.

그런데 then은 비동기 처리 함수에서 발생한 에러만 캐치하지만 catch는 then에서 발생한 에러도 캐치하므로 catch를 사용하는 편이 더 효율적.


## 6. Promise Chaining

비동기 함수 처리 결과로 다른 비동기 함수를 호출해야하는 경우 프로미스는 후속 처리 메소드를 체이닝(chaining)해서 여러 개의 프로미스를 연결해서 사용할 수 있다. 이로써 콜백 헬을 해결한다.

Promise 객체를 반환한 비동기 함수는 프로미스 후속 처리 메소드인 then이나 catch를 사용할 수 있다. 따라서 then 메소드가 Promise 객체를 반환하도록 하면 여러 개의 프로미스를 연결해 사용할 수 있다.

```javascript
<!DOCTYPE html>
<html>
<body>
  <pre class="result"></pre>
  <script>
    const $result = document.querySelector('.result');
    const render = content => { $result.textContent = JSON.stringify(content, null, 2); };

    const promiseAjax = (method, url, payload) => {
       // Promise를 반환하는 Ajax 함수
      return new Promise((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        xhr.open(method, url);
        xhr.setRequestHeader('Content-type', 'application/json');
        xhr.send(JSON.stringify(payload));

        xhr.onreadystatechange = function () {
          if (xhr.readyState !== XMLHttpRequest.DONE) return;

          if (xhr.status >= 200 && xhr.status < 400) {
            resolve(xhr.response); // Success!
          } else {
            reject(new Error(xhr.status)); // Failed...
          }
        };
      });
    };

    const url = 'http://jsonplaceholder.typicode.com/posts';

    // 포스트 id가 1인 포스트를 검색하고 프로미스를 반환한다.
    promiseAjax('GET', `${url}/1`)
      // 포스트 id가 1인 포스트를 작성한 사용자의 아이디로 작성된 모든 포스트를 검색하고 프로미스를 반환한다.
      .then(res => promiseAjax('GET', `${url}?userId=${JSON.parse(res).userId}`))
      .then(JSON.parse)
      .then(render)
      .catch(console.error);
  </script>
</body>
</html>
```



## 7. Promise의 정적 메소드

### 7.1. Promise.resolve / Promise.reject

Promise.resolve와 Promise.reject는 존재하는 값을 Promise로 래핑하기 위해서 사용

Promise.resolve는 인자로 전달된 값을 resolve하는 Promise를 생성

```javascript
const resolvedPromise = Promise.resolve([1, 2, 3]);
resolvedPromise.then(console.log); // [ 1, 2, 3 ]
```

```javascript
// 위와 동일하게 동작
const resolvedPromise = new Promise(resolve => resolve([1, 2, 3]));
resolvedPromise.then(console.log); // [ 1, 2, 3 ]
```

Peromise.reject도 인자로 전달된 값을 reject하는 프로미스를 생성한다.

```javascript
const rejectedPromise = Promise.reject(new Error('Error!'));
rejectedPromise.catch(console.log); // Error: Error!
```

```javascript
// 위와 동일하게 동작
const rejectedPromise = new Promise((resolve, reject) => reject(new Error('Error!')));
rejectedPromise.catch(console.log); // Error: Error!
```

### 7.2. Promise.all

Promise.all 메소드는 Promise가 담겨있는 배열 등의 iterable을 인자로 전달 받음

그리고 **전달 받은 모든 Promise를 병렬로 처리하고 처리 결과를 resolve하는 새로운 Promise를 반환**

```javascript
Promise.all([
  new Promise(resolve => setTimeout(() => resolve(1), 3000)), // 1
  new Promise(resolve => setTimeout(() => resolve(2), 2000)), // 2
  new Promise(resolve => setTimeout(() => resolve(3), 1000))  // 3
]).then(console.log) // [ 1, 2, 3 ]
  .catch(console.log);
```

- 첫 번째 Promise는 3초 후에 1을 resolve해서 처리 결과를 반환
- 두 번째 Promise는 2초 후에 2를 resolve해서 처리 결과를 반환
- 세 번째 Promise는 1초 후에 3을 resolve해서 처리 결과를 반환

Promise.all 메소드는 전달받은 모든 Promise를 병렬로 처리한다. 이때 모**든 Promise의 처리가 종료될 때까지 기다린 후** 아래와 같이 처리 결과를 resolve / reject 한다.

- 모든 Promise가 성공하면 각각 Promise가 resolve한 처리 결과를 배열에 담아 resolve하는 새로운 Promise를 반환. 첫번째 Promise가 가장 나중에 처리돼도 전달 받은 Promise의 순서대로 반환. 즉, **처리 순서가 보장됨**
- Promise의 처리가 하나라도 실패하면 가장 먼저 실패한 Promise가 reject한 에러를 reject하는 새로운 Promise를 즉시 반환



### 7.3. Promise.race

Promise.race 메소드는 Promise.all 메소드처럼 모든 프로미스를 병렬 처리하는 것이 아니라 **가장 먼저 처리된 프로미스가 resolve한 처리 결과를 resolve하는 새로운 프로미스를 반환**한다.

```javascript
Promise.race([
  new Promise(resolve => setTimeout(() => resolve(1), 3000)), // 1
  new Promise(resolve => setTimeout(() => resolve(2), 2000)), // 2
  new Promise(resolve => setTimeout(() => resolve(3), 1000))  // 3
]).then(console.log) // 3
  .catch(console.log);
```



# async/await 